breacharbiter.go

```diff
@@ -533,40 +510,55 @@ func (b *breachArbiter) exactRetribution(
 	brarLog.Debugf("Breach transaction %v has been confirmed, sweeping "+
 		"revoked funds", breachInfo.commitHash)
 
-	// With the breach transaction confirmed, we now create the justice tx
-	// which will claim ALL the funds within the channel.
-	justiceTx, err := b.createJusticeTx(breachInfo)
+	finalTx, err := b.cfg.Store.GetFinalizedTxn(&breachInfo.chanPoint)
 	if err != nil {
-		brarLog.Errorf("unable to create justice tx: %v", err)
+		brarLog.Errorf("unable to get finalized txn for"+
+			"chanid=%v: %v", &breachInfo.chanPoint, err)
 		return
 	}
 
+	// If this retribution has not been finalized before, we will first
+	// construct a sweep transaction and write it to disk. This will allow
+	// the breach arbiter to re-register for notifications for the justice
+	// txid.
+	if finalTx == nil {
+		// With the breach transaction confirmed, we now create the
+		// justice tx which will claim ALL the funds within the channel.
+		finalTx, err = b.createJusticeTx(breachInfo)
+		if err != nil {
+			brarLog.Errorf("unable to create justice tx: %v", err)
+			return
+		}
+
+		// Persist our finalized justice transaction before making an
+		// attempt to broadcast.
+		err := b.cfg.Store.Finalize(&breachInfo.chanPoint, finalTx)
+		if err != nil {
+			brarLog.Errorf("unable to finalize justice tx for "+
+				"chanid=%v: %v", &breachInfo.chanPoint, err)
+			return
+		}
+	}
+
 	brarLog.Debugf("Broadcasting justice tx: %v",
 		newLogClosure(func() string {
-			return spew.Sdump(justiceTx)
+			return spew.Sdump(finalTx)
 		}))
 
-	_, currentHeight, err := b.cfg.ChainIO.GetBestBlock()
-	if err != nil {
-		brarLog.Errorf("unable to get current height: %v", err)
-		return
-	}
-
 	// Finally, broadcast the transaction, finalizing the channels'
 	// retribution against the cheating counterparty.
-	if err := b.cfg.PublishTransaction(justiceTx); err != nil {
+	if err := b.cfg.PublishTransaction(finalTx); err != nil {
 		brarLog.Errorf("unable to broadcast "+
 			"justice tx: %v", err)
-		return
```

Comment:
this assumes that all failures from publish txn are duplicates like in the nursery. even if it did fail to publish, there's no real harm in setting up the remainder of this section since it only is activated if the tx confirms. this should later be replaced with the concrete error types